internal/component/all/all_test.go

```diff
@@ -0,0 +1,175 @@
+		require.True(t, ok, "Expected component %q to exist", componentName)
+
+		t.Run(reg.Name, func(t *testing.T) {
+			testNoReusePointer(t, reg)
+		})
+	}
+}
+
+func testNoReusePointer(t *testing.T, reg component.Registration) {
+	t.Helper()
+
+	var (
+		args1 = reg.CloneArguments()
+		args2 = reg.CloneArguments()
+	)
+
+	if args1, ok := args1.(river.Defaulter); ok {
+		args1.SetToDefault()
+	}
+	if args2, ok := args2.(river.Defaulter); ok {
+		args2.SetToDefault()
+	}
+
+	rv1, rv2 := reflect.ValueOf(args1), reflect.ValueOf(args2)
+	ty := rv1.Type().Elem()
+
+	// Edge case: if the component's arguments type is an empty struct, skip.
+	// Not skipping causes the test to fail, due to an optimization in
+	// reflect.New where initializing the same zero-length object results in the
+	// same pointer.
+	if rv1.Elem().NumField() == 0 {
+		return
+	}
+
+	if path, shared := sharePointer(rv1, rv2); shared {
+		fullPath := fmt.Sprintf("%s.%s.%s", ty.PkgPath(), ty.Name(), path)
+
+		assert.Fail(t,
+			fmt.Sprintf("Detected SetToDefault pointer reuse at %s", fullPath),
+			"Types implementing river.Defaulter must not reuse pointers across multiple calls. Doing so leads to default values being changed when unmarshaling configuration files. If you're seeing this error, check the path above and ensure that copies are being made of any pointers in all instances of SetToDefault calls where that field is used.",
+		)
+	}
+}
+
+func sharePointer(a, b reflect.Value) (string, bool) {
+	// We want to recursively check a and b, so if they're nil they need to be
+	// initialized to see if any of their inner values have shared pointers after
+	// being initialized with defaults.
+	initValue(a)
+	initValue(b)
+
+	// From the documentation of reflect.Value.Pointer, values of chan, func,
+	// map, pointer, slice, and unsafe pointer are all pointer values.
+	//
+	// Additionally, we want to recurse into values (even if they don't have
+	// addresses) to see if there's shared pointers inside of them.
+	switch a.Kind() {
+	case reflect.Chan, reflect.Func, reflect.UnsafePointer:
+		return "", a.Pointer() == b.Pointer()
+
+	case reflect.Map:
+		if pointersMatch(a, b) {
+			return "", true
+		}
+
+		iter := a.MapRange()
+		for iter.Next() {
+			aValue, bValue := iter.Value(), b.MapIndex(iter.Key())
+			if !bValue.IsValid() {
+				continue
+			}
+			if path, shared := sharePointer(aValue, bValue); shared {
+				return path, true
+			}
+		}
+		return "", false
+
+	case reflect.Pointer:
+		if pointersMatch(a, b) {
+			return "", true
+		} else {
+			// Recursively navigate inside of the pointer.
+			return sharePointer(a.Elem(), b.Elem())
+		}
+
+	case reflect.Interface:
+		if a.UnsafeAddr() == b.UnsafeAddr() {
+			return "", true
+		}
+		return sharePointer(a.Elem(), b.Elem())
+
+	case reflect.Slice:
+		if pointersMatch(a, b) {
+			// If the slices are nil, we don't need to check for shared pointers
+			if a.Len() > 0 || b.Len() > 0 {
+				return "", true
+			}
```

Comment:
This isn't quite right, since it could be sharing a pointer where len=0 but cap>0, which would be an error. This playground link shows that the pointer for slices is fixed until either len>0 or cap>0: https://go.dev/play/p/vUsVSJ-Z9jo

So this condition should be:

```go 
if a.Len() > 0 || a.Cap() > 0 || b.Len() > 0 || b.Cap() > 0 {
```

Also note that that the comment here isn't quite right. `[]string{}` is not nil, it's just a preallocated immutable pointer, so it's safe for us to ignore.

(This longer condition may be worth a helper function so it can be explained)